Switch to software if NX is spending too much time in a job

rzinsly opened the pull request:
Measure the average time spent on NX jobs and fallback to software if it is greater than our threshold.
I'm keeping it as draft for now because I'm not sure which value should be used as threshold yet.
This seems to solve #152 as well.

tuliom commented on lib/nx_zlib.h:
```diff
@@ -336,6 +338,45 @@ typedef struct nx_stream_s {
+static const uint64_t decay = 4;
+
+/* For estimating device throughput */
+static inline int nx_device_stats(uint64_t start, uint64_t end)
```
Why is the implementation of this function in a header?

tuliom commented on lib/nx_zlib.h:
```diff
@@ -336,6 +338,45 @@ typedef struct nx_stream_s {
+	delay = (last_dly + decay * delay) / (decay + 1);
+
+	/* this is the job delay through this device */
+	avg_delay = delay;
```
I'm not sure this implementation works well for multi-thread.
There will be lost values if both try to write `avg_delay` at the same time.

Is it OK to lose measurements?
I'm afraid we need at least a relaxed atomic write to avg_delay in order to guarantee its integrity.

tuliom commented on lib/nx_zlib.h:
```diff
@@ -336,6 +338,45 @@ typedef struct nx_stream_s {
+	if (last_dly > nx_ticks_per_sec ||
+	    (last_dly < nx_ticks_per_sec/10000000UL)) {
+		prt_err("Unrealiable delay: %lu tps: %lu\n", last_dly, nx_ticks_per_sec);
+		return -1; /* unreliable */
```
This function is called in 2 places, but none of them check for the return.
Does it really need to return an integer?

rzinsly replied:
Right, I will remove this.

tuliom commented on test/libnxz.abi:
```diff
@@ -80,7 +80,9 @@
   <elf-variable-symbols>
+    <elf-symbol name='avg_delay' size='8' type='object-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
     <elf-symbol name='is_deflate' size='1' type='object-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='metric' size='8' type='object-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
```
Shouldn't they be hidden? i.e. defined with `__attribute__ (visibility ("hidden")`)

tuliom commented on lib/nx_zlib.h:
```diff
@@ -336,6 +338,45 @@ typedef struct nx_stream_s {
 typedef struct nx_stream_s *nx_streamp;
 
+/* metric used to decide if it's time to switch to sw, as we are not able
+   to switch mid-stream we cannot use avg_delay directly.  */
```
I didn't understand this comment. I'm afraid I'm not understanding because I can't distinguish how `metric` and `avg_delay` are being used.
What prevents `metric` from being updated after every `nx_submit_job()`?

rzinsly replied:
`metric` is used for the comparison in the `use_nx_*` functions, if we update it on `nx_submit_job()` and it passes the threshold it would try to switch to software in the middle of a NX stream and it would fail.

tuliom replied:
Then I'm afraid we have an issue here because other threads may update `metric` causing a thread to try to switch in the middle of the execution of a stream.

Should `use_nx_*` only be called by `*Init` functions?

tuliom commented on lib/nx_zlib.h:
```diff
@@ -354,12 +395,15 @@ static inline int use_nx_inflate(uLong len)
+	/* #1 Time spent */
+	if(metric > METRIC_THRESHOLD) return 0;
+
+	/* #2 Treshold */
```
Typo: Threshold

tuliom commented on lib/nx_zlib.h:
```diff
@@ -92,6 +92,8 @@ struct selector {
 #define COMPRESS_THRESHOLD	(1024)
 #define DECOMPRESS_THRESHOLD	(1024)
 
+#define METRIC_THRESHOLD	8000
```
I wonder if this should be configurable....

tuliom reviewed (changes requested):
The PR looks mostly good to me, but I have a few questions and I think there are a few mistakes.
More source code comments might be needed too.

In the first commit message: `s/changing they/changing them/`

ThinkOpenly commented on lib/nx_deflate.c:
```diff
@@ -704,6 +704,9 @@ int nx_deflateInit2_(z_streamp strm, int level, int method, int windowBits,
 
+	s->sw_stream = NULL;
+	s->switchable = 0;
+	s->use_nx = avg_delay > nx_config.delay_threshold ? 0 : 1;
```
nit: you could say:
```
s->use_nx = avg_delay <= nx_config.delay_threshold
```

rzinsly replied:
Done, thanks!

tuliom commented on lib/nx_gzlib.c:
```diff
@@ -244,10 +244,10 @@ int nx_gzwrite (gzFile file, const void *buf, unsigned len)
+	z_streamp stream = &(state->strm);
 
-	if (sel == GZIP_SW)
+	if (0 == has_nx_state(stream))
```
In PR #169 , I'm proposing to rewrite these gz* functions and reuse the code from deflate*/inflate* family of functions.
I believe one of the patches there helps to simplify this part here.
If there is an agreement on what I proposed there, I suspect this commit could be removed.

rzinsly replied:
Ok, I'm keeping the commit for now.

tuliom reviewed (approved):
LGTM.
I'm sorry, but a recent merge to branch develop caused a conflict with this PR.
There is another PR conflicting with the changes here.

rzinsly commented:
I rebased on top of develop and added code to decrease avg_delay when we choose software, otherwise after the switch we stop measuring time and never go back to NX.

rzinsly commented:
I split the threshold values between compression and decompression as decompression with software is faster. Also added a multiplier for baremetal as the latency is higher there.
I choose higher thresholds trying to be on the safe side and just avoid starvation instead of try to get the best performance.
The PR is open for review but I'll wait for the resolution of #169 before merging it.

rzinsly commented:
I rebased the code and removed commit 621bac4.
